src/semaphore.rs

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2021 Two Sigma Open Source, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// clippy does a false alarm if you're trying to use a Condvar.
+//
+// https://rust-lang.github.io/rust-clippy/master/index.html#mutex_atomic
+// https://github.com/rust-lang/rust-clippy/issues/1516
+#![allow(clippy::mutex_atomic)]
+
+use std::sync::{Arc, Condvar, Mutex};
+use std::time::Duration;
+
+/// A counting semaphore.
```

Comment:
Is there no semaphore crate we can just use? I'm a little surprised we need to roll our own here.

Reply:
The ones that [show up on crates.io](https://crates.io/search?q=semaphore) look like they're really low usage. This felt better than taking a dependency on something unowned/unmaintained.